### DISCUSSION TITLE
nextcloud: program escapes dbus-proxy via systemd

### DISCUSSION THREAD

**Discussion by michelesr** [netblue30/firejail#6209] [Q&A] [answered]
I've started seeing this weird behavior in Nextcloud where it uses dbus to spawn a process via systemd.

```
systemctl --user -a  | rg nextcl
  dbus-:1.2-com.nextcloudgmbh.Nextcloud@0.service                                                                                                       loaded    active   running   dbus-:1.2-com.nextcloudgmbh.Nextcloud@0.service
  app-dbus\x2d:1.2\x2dcom.nextcloudgmbh.Nextcloud.slice                                                                                                 loaded    active   active    Slice /app/dbus-:1.2-com.nextcloudgmbh.Nextcloud
```
  
Seems weird that Nextcloud can do that with the default firejail profile that has `dbus-user filter` and `dbus-user.talk org.freedesktop.secrets` only.
  
Could it be that dbus-proxy is not effective anymore now that Arch Linux has switched to [dbus-broker](https://github.com/bus1/dbus-broker)?

Relates to:

* #6211

**Comment by michelesr** [comment, marked as answer]
Actually... I've tested dbus-proxy and it works fine. It's probably something else triggering the systemd service over dbus.

```
cat /usr/share/dbus-1/services/com.nextcloudgmbh.Nextcloud.service
[D-BUS Service]
Name=com.nextcloudgmbh.Nextcloud
Exec=/usr/bin/nextcloud --background
```

**Reply by michelesr** [reply]
Yes, it seems to be Nautilus trying to communicate with Nextcloud trough dbus and triggering the systemd service to be activated.

**Reply by michelesr** [reply]
You can add `dbus-user.own com.nextcloudgmbh.Nextcloud` to the profile so that the sandboxed Nextcloud can answer on that name so that the broker won't ask systemd to spawn a non-sandboxed Nextcloud process, or you could set up dbus-filtering for Nautilus, depending on your hardening requirements.

**Reply by ghost** [reply]
> Exec=/usr/bin/nextcloud --background

You'll need to create a `local` copy of that com.nextcloudgmbh.Nextcloud.service file and make it use the firejailed nextcloud instance created via firecfg. This is quite common for systemd services that point directly at the non-sandboxed application.

```console
$ cat  ~/.local/share/dbus-1/services/com.nextcloudgmbh.Nextcloud.service
[D-BUS Service]
Name=com.nextcloudgmbh.Nextcloud
Exec=/usr/local/bin/nextcloud --background
```

After doing so you can add the below to let the sandboxed version own that D-Bus address:

```console
$ cat  ~/.config/firejail/nextcloud.local
dbus-user.own com.nextcloudgmbh.Nextcloud
```

Probably wise to reload the systemd user daemon afterwards. Please let us know how it all works.

**Reply by michelesr** [reply]
> You'll need to create a local copy of that com.nextcloudgmbh.Nextcloud.service

Yes, that worked, although `systemctl --user daemon-reload` wasn't enough, so I rebooted the machine.

**Reply by kmk3** [reply]
@glitsj16 [on Feb 15](https://github.com/netblue30/firejail/discussions/6209#discussioncomment-8486688):

> ```
> $ cat  ~/.config/firejail/nextcloud.local
> dbus-user.own com.nextcloudgmbh.Nextcloud
> ```

Should this be added to nextcloud.profile?

**Reply by ghost** [reply]
> Should this be added to nextcloud.profile?

@kmk3

Yes we should. But that still leaves the problem that the current firecfg implementation doesn't create an override in ~/.local/share/dbus-1/services. That's a feature which only @rusty-snake's [firecfg.py](https://github.com/rusty-snake/firecfg.py) automates AFAIK. Not a very elegant situation IMO if we need to add a convoluted comment in nextcloud.profile to explain all this...

Any thoughts?

**Reply by kmk3** [reply]
> Yes we should. But that still leaves the problem that the current firecfg
> implementation doesn't create an override in ~/.local/share/dbus-1/services.
> That's a feature which only @rusty-snake's
> [firecfg.py](https://github.com/rusty-snake/firecfg.py) automates AFAIK. Not
> a very elegant situation IMO if we need to add a convoluted comment in
> nextcloud.profile to explain all this...
> 
> Any thoughts?

I'm not familiar with dbus services, but if it's similar to the overrides of
.desktop files, then it makes sense to me.

**Reply by ghost** [reply]
> I'm not familiar with dbus services, but if it's similar to the overrides of
.desktop files, then it makes sense to me.

That's correct, this is indeed a similar situation. I'm preparing an entree on D-Bus overrides for the wiki. The current FAQ on [firecfg](https://github.com/netblue30/firejail/wiki/Frequently-Asked-Questions#how-do-i-undo-firecfg) needs updating anyway now there's native support via `firecfg.d`. When that's ready I can point the nextcloud comment to it. Should make it a bit easier to keep the comment within reasonable limits.

**Reply by michelesr** [reply]
Even if you don't add the dbus service override, most users will start Nextcloud at login anyway, and if it owns the dbus name then it will respond to it, and the broker won't need to spawn another instance of Nextcloud. The problem will then arise when the firejailed Nextcloud won't be running, but most of the users will probably keep it running all the time in background anyway.
